### PR TITLE
rfb: VeNCrypt Plain SecurityType support

### PR DESCRIPTION
**Pull request by regnarg**
This allows using TigerVNC server with PAM authentication (e.g. against
LDAP or other extensible authentication mechanisms)

Tested with TigerVNC server (Xvnc -SecurityTypes Plain -PlainUsers '*')

Should not break anything else, this method is tried last when all
other fail.

Tested in Firefox 74 and Chromium 80

### CODE REVIEW

**Comment by regnarg** [comment]
NB: I have no idea why CI check fails, does not seem related to my changes.

**Comment by regnarg** [comment]
I tried to improve the formatting and add blank lines for clarity.

File: core/rfb.js [hunk 1]
```diff
@@ -1018,6 +1021,93 @@ export default class RFB extends EventTargetMixin {
...
+    // protocol: https://github.com/rfbproto/rfbproto/blob/617d72b4ced0da7d90d4a70cc6d6e1a30819f597/rfbproto.rst#vencrypt
+    _negotiate_vencrypt_auth() {
+
+        if (!this._rfb_credentials.username ||
+            !this._rfb_credentials.password) {
+            this.dispatchEvent(new CustomEvent(
+                "credentialsrequired",
+                { detail: { types: ["username", "password", "target"] } }));
+            return false;
+        }
```
Comment (CendioOssman) [inline core/rfb.js]: You don't know that you'll need this yet. The server might not support Plain.

File: core/rfb.js [hunk 2]
```diff
@@ -1018,6 +1021,93 @@ export default class RFB extends EventTargetMixin {
...
+                // 0x100 = 256
+                this._sock.send([0, 0, 1, 0]);
+                const user = this._rfb_credentials.username;
+                const pass = this._rfb_credentials.password;
+
+                // XXX we assume lengths are <= 255 (should not be an issue in the real world)
+                this._sock.send([0, 0, 0, user.length]);
+                this._sock.send([0, 0, 0, pass.length]);
+                this._sock.send_string(user);
+                this._sock.send_string(pass);
```
Comment (CendioOssman) [inline core/rfb.js]: These two need some encoding here. Most likely UTF-8.

File: core/rfb.js [hunk 3]
```diff
@@ -1018,6 +1021,93 @@ export default class RFB extends EventTargetMixin {
...
+        }
+
+        // waiting for subtypes list
+        if (this._rfb_vencrypt_state == 3) {
+            if (this._sock.rQwait("vencrypt subtypes", 4 * this._rfb_vencrypt_subtypes_length)) { return false; }
+
+            const subtypes = [];
+            for (let i = 0; i < this._rfb_vencrypt_subtypes_length; i++) {
+                subtypes.push(this._sock.rQshift32());
+            }
```
Comment (CendioOssman) [inline core/rfb.js]: Can we restructure things so this is shared with the normal handling? VeNCrypt is a super set of the normal authentication after all.

File: core/rfb.js [hunk 4]
```diff
@@ -1018,6 +1021,93 @@ export default class RFB extends EventTargetMixin {
         return this._negotiate_authentication();
     }
 
+    // VeNCrypt authentication, currently only supports version 0.2 and only Plain subtype
+    // protocol: https://github.com/rfbproto/rfbproto/blob/617d72b4ced0da7d90d4a70cc6d6e1a30819f597/rfbproto.rst#vencrypt
```
Comment (CendioOssman) [inline core/rfb.js]: We don't need this as all of this file uses that specification as the reference.

**Comment by regnarg** [comment]
> Out of curiosity, what does your setup look like?
I use TigerVNC with Plain auth against an LDAP server using pam_ldap.

> I figured Plain was rarely used in TigerVNC as it requires running the server as root.
Only when authenticating against local user database (PAM needs root to access /etc/shadow). With LDAP you can do it without root, pam_ldap just verifies the credentials with LDAP server.

File: core/rfb.js (hunk 3 above)
Comment (regnarg) [reply core/rfb.js]: I am not sure how one would do that. VeNCrypt is one security type, which has subtypes. You first negotiate the VeNCcrypt security type and then negotiate the subtype. The subtype negotiation protocol is different from the type negotiation protocol. For example, subtype numbers are 32 bit but type numbers are 8 bit.

**Comment by samhed** [comment]
@regnarg do you think you will get a chance to make the requested changes?

**Comment by regnarg** [comment]
> @regnarg do you think you will get a chance to make the requested changes?

Yes, I will update it soon.

**Comment by regnarg** [comment]
Updated according to requested changes.

**Comment by regnarg** [comment]
On Tue, Apr 07, 2020 at 11:55:33PM -0700, Pierre Ossman (Work account) wrote:
> LGTM too. Could you squash the commits in to one?

Squashed.

**Pull request merged**